README.md
# Kyle's Peripheral Abstraction Layer 

[![CircleCI](https://circleci.com/gh/kmdouglass/kpal.svg?style=svg)](https://circleci.com/gh/kmdouglass/kpal)
[![Docs.rs](https://docs.rs/kpal/badge.svg)](https://docs.rs/kpal)
[![Crates.io](https://img.shields.io/crates/v/kpal)](https://crates.io/crates/kpal)

KPAL is an extensible control system for physical computing.

## Documentation

**KPAL is under development. The API will not be considered stable until 1.0 is released.**

- [kpal-plugin](https://docs.rs/kpal-plugin/0.1.0/kpal_plugin/) - Used to write plugins for KPAL
- [kpal-gpio-cdev](https://docs.rs/kpal-gpio-cdev/0.1.0/kpal_gpio_cdev/) - Controls the GPIO pins
  on a Raspberry Pi

## Overview

KPAL allows you to control and read data from peripherals attached to a computer such as your
desktop or Raspberry Pi. It acts as an interface between users and individual peripherals through
two application programming interfaces (APIs):

- **the user API** A web service that can be accessed from different computers on a network,
  including smart phones
- **the plugin API** A high-level plugin interface that allows KPAL to communicate with
  peripherals such as sensors, motors, and cameras

## Quickstart

1. Download the archive that matches the latest version of the binaries for your platform from the
   [releases page](https://github.com/kmdouglass/kpal/releases).
2. Unpack the archive.
3. Create the following folder in your home directory:

```console
mkdir -p ~/.kpal/libraries
```
4. Move the file `libbasic-plugin.so` from the archive into the `~/.kpal/libraries` folder. This is
   an example plugin that is used for demonstrations and testing; it does not control any actual
   hardware.
5. Run the binary file `kpald` to start the daemon. If you want to see the logs, set the `RUST_LOG`
   environment variable to `info`, `error`, or `debug`, depending on the desired log level:

```console
RUST_LOG=info ./kpald
```

You may now make HTTP requests to the daemon. The following examples use the UNIX `curl` command
line utility to make the requests, but you may use the HTTP client of your choice.

```console
# Get the libraries that are available to the daemon
curl -s localhost:8000/api/v0/libraries

# Get the library with ID 0
curl -s localhost:8000/api/v0/libraries/0

# Create a new peripheral from the library with ID 0
curl -s \
     --request POST \
     localhost:8000/api/v0/peripherals \
     --header "Content-Type: application/json" \
     --data '{"name":"foo","library_id":0}'

# Create a new peripheral and override the default value of a pre-init attribute
curl -s \
     --request POST \
     localhost:8000/api/v0/peripherals \
     --header "Content-Type: application/json" \
     --data '{
         "name": "foo",
         "library_id": 0,
         "attributes": [
             {"id":0, "type":"double", "value": 999.99}
         ]
     }'

# Get all the peripherals currently managed by the daemon
curl -s localhost:8000/api/v0/peripherals

# Get the peripheral with ID 0
curl -s localhost:8000/api/v0/peripherals/0

# Get the attributes of the peripheral with ID 0
curl -s localhost:8000/api/v0/peripherals/0/attributes

# Get the attribute with ID 0 from the peripheral with ID 0
curl -s localhost:8000/api/v0/peripherals/0/attributes/0

# Set the value of the attribute with ID 0 of the peripheral with ID 0
curl -s \
     --request PATCH \
     localhost:8000/api/v0/peripherals/0/attributes/0 \
     --header "Content-Type: application/json" \
     --data '{"type":"double","value":42}'
```

## Core components

 
                +--------------------------------+
                |                                |
                |          Object Model          |
                |                                |      ^
                +--------------------------------+      |
             ----------------------------------------   |  User API
                +--------------------------------+      |
                |                                |      v
                |        REST Integration        |
                |                                |
                +--------------------------------+
             ----------------------------------------
                +--------------------------------+
                |                                |
                |           KPAL Daemon          |
                |                                |      ^
                +--------------------------------+      |
             ----------------------------------------   |  Plugin API
                +--------+  +--------+  +--------+      |
                |        |  |        |  |        |      v
                | Plugin |  | Plugin |  | Plugin |
                |        |  |        |  |        |
                +--------+  +--------+  +--------+


### Object model

The object model is the set of resources with which users interact. Currently, these resources
include:

- **peripherals** Models of individual hardware peripherals
- **attributes** Attributes describe a peripheral. Users interact with peripherals by modifying or
  reading their attributes.
- **libraries** Shared libraries provide a unique implementation of the plugin interface for each
  component.

### Integrations

Integrations are different implementations of the user API. Examples of possible integrations
include

 - a JSON REST API
 - gRPC
 - a C static library
 - Python bindings

Currently only a JSON REST integration is available, but the structure of KPAL makes it relatively
easy to add others.

### Daemon

The KPAL daemon, or `kpald`, is a server that runs on the computer to which the peripherals are
connected. Users directly interact with the daemon through the user API. Each peripheral runs
inside its own thread which is spawned by a POST request to the user API. The daemon forwards other
user requests to each thread through the thread's dedicated channel. The threads interpret the
incoming requests and, in response, read and write data to individual plugins through the plugin
API using shared libraries.

### Plugins

Plugins are the means by which peripherals are integrated into KPAL. A plugin uses a shared library
(a `.so` file on Linux) to communicate with the daemon. The common set of functions that the
library provides is the plugin API. Any programming language that can provide a C language
interface can be used to write a plugin library.

A plugin combines the data that represents a peripheral's state with the functionality for
controlling the hardware device that is modeled by the peripheral.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] - 2020-03-29
### Added
- A new error code was added to the `kpal-plugin` crate called `STRING_CONVERSION_ERR`.
- A new module called `integrations` was added. Its purpose is to make it easier to create
  different types of user APIs and to decouple the KPAL models from their input/output
  representations.
- A crate-level docstring was added to the `kpal` crate.

### Changed
- The integration tests were refactored and expanded to cover more use cases.
- JSON, not text, is now returned in HTTP error responses.
- The internal handling of matches on pairs of enums was redone. This has no effect on users, but
  for developers the compiler should now be able to remind you when you forget to update methods
  involving matches on pairs after introducing a new enum variant for `Attribute`, `Value`, etc.
- Error handling in `kpal-plugin` and `kpal-gpio-cdev` was refactored to condense the number of
  error types exported by `kpal-plugin` to one.
  
### Fixed
- `kpal-gpio-cdev` now returns the correct `PLUGIN_INIT_ERR` error code when the plugin fails to
  initialize. Previously it was returning `NUMERIC_CONVERSION_ERR`.

## [0.2.2] - 2020-03-07
### Changed
- Module structures were rerranged so that module APIs are no longer nested; there is only layer of
  modules below the KPAL root layer of modules.
- All modules, including those hidden inside the public ones, now export only one Error type.

### Fixed
- `kpal` and `kpal-gpio-cdev` now depend on the correct version of `kpal-plugin`.

## [0.2.1] - 2020-03-01
### Changed
- The method to create new peripherals now accepts a JSON object and not an array for the pre-init
  attributes.
- The `links` field of a peripheral JSON resource is now renamed to `attributes`. Its name in the
  Rust code remains unchanged.

### Fixed
- A plugin executor can no longer get stuck in an infinite loop when initializing if a plugin
  provides no attributes. This was fixed by adding two new methods to the plugin API:
  `attribute_count` and `attribute_ids`.
- A delay was added between publishing jobs on the CI so that crates.io would have time to update
  before subsequent crates were published.

## [0.2.0] - 2020-02-09
### Added
- A Dockerfile was added at
  [resources/docker/rust-linux-x86_64/Dockerfile](resources/docker/rust-linux-x86_64/Dockerfile)
  that defines the image in which the x86_64 CI jobs are run.
- A new workflow was added to the CI configuration for releases. The new workflow is called `main`
  and has two pathways. One is triggered by a commit and one by pushing a tag to the Git repo. The
  pathway that is triggered by a commit contains a subset of the jobs contained by the one that is
  triggered by tag. This enables a release-from-tag strategy.
- A build cache was added to the CI that includes the `~/.cargo` and `target` folders.
- A job was added to the CI to build artifacts for the ARMv7 architecture (the one used by the
  Raspberry Pi). The Dockerfiles for the builder and tester are located
  [here](resources/docker/rust_cross_armv7-linux-x86_64/Dockerfile) and
  [here](resources/docker/kpal_tester-linux-armv7/Dockerfile), respectively.
- Added the Dependabot dependencey management service to the GitHub repo.
- Added steps to the CI to check that the code has been linted with `cargo clippy`.
- A new set of jobs was added to the CI to build and publish Dockerfiles to Docker Hub.
- `kpal-plugin` now exposes a `KpalLibraryInit` type that is used by the daemon to initialize
  plugin libraries. Previously, the function signature was hard-coded.
- A new macro called `declare_plugin` was added to the `kpal-plugin` library. This macro takes care
  of initializing the FFI code for plugins so that plugin authors do not have to.
- Callbacks were added to the `kpal-plugin` library, making it easier to isolate the parts of the
  plugin code that communicate with the hardware.
- Peripherals may how have string attributes that contain any character within the ASCII character
  set except for the null byte.
- Peripherals now have `pre-init` attributes. These allow you to set attribute values before the
  plugin is initialized, which improves the ergonomics of writing plugins.
- A new unsigned integer Attribute type was added.

### Changed
- All artifacts are now built on the CI with the `--release` profile.
- The entire KPAL codebase is now linted with `clippy`.
- The `plugins::driver` and `plugins::init` modules were moved into methods of the Executor struct
  that is provided by the `plugins::executor` module.
- All errors in the `plugins` module were consolidated and moved into a `plugins::errors`
  submodule. Likewise, all errors in the `web` module were consolidated and moved into a
  `web::errors` submodule.
- The `Peripheral` in the `kpal-plugin` crate was renamed to `PluginData` to avoid confusion with
  the `Peripheral` of the user API and to emphasize that it holds the state of a plugin.
- Many methods in the `PluginAPI` trait now have default implementations so that they are no longer
  required in a plugin library.
- Changed the name of the `kpal_plugin_init` FFI function to `kpal_plugin_new`.

### Fixed
- The integration and unit tests no longer look in only the `target/debug` directory for test
  artifacts. Instead, they search for artifacts in the parent folder and subfolders of the
  currently-running test binary.
- Fixed a unit test that was not compiling on 32-bit Linux platforms due to a difference in integer
  size as compared to 64-bit systems.
- The description of kpald in the --help text now matches the one on GitHub and in Cargo.toml.

## [0.1.0] - 2019-12-14
### Added
- A `CHANGELOG.md` file to track project changes.

src/lib.rs
//! KPAL is an extensible control system for physical computing.
//!
//! # Overview
//!
//! KPAL allows you to control and read data from peripherals attached to a computer such as your
//! desktop or Raspberry Pi. It acts as an interface between users and individual peripherals
//! through two application programming interfaces (APIs):
//!
//! - **the user API** A service that can be accessed from different computers on a network,
//! including smart phones
//! - **the plugin API** A high-level plugin interface that allows KPAL to communicate with
//! peripherals such as sensors, motors, and cameras
//!
//! # Quickstart
//!
//! 1. Download the archive that matches the latest version of the binaries for your platform from
//!    the [releases page](https://github.com/kmdouglass/kpal/releases).
//! 2. Unpack the archive.
//! 3. Create the following folder in your home directory:
//!
//! ```console
//! mkdir -p ~/.kpal/libraries
//! ```
//! 4. Move the file `libbasic-plugin.so` from the archive into the `~/.kpal/libraries`
//!    folder. This is an example plugin that is used for demonstrations and testing; it does not
//!    control any actual hardware.
//! 5. Run the binary file `kpald` to start the daemon. If you want to see the logs, set the
//! `RUST_LOG` environment variable to `info`, `error`, or `debug`, depending on the desired log
//! level:
//!
//! ```console
//! RUST_LOG=info ./kpald
//! ```
//!
//! You may now make HTTP requests to the daemon. The following examples use the UNIX `curl`
//! command line utility to make the requests, but you may use the HTTP client of your choice.
//!
//! ```console
//! # Get the libraries that are available to the daemon
//! curl -s localhost:8000/api/v0/libraries
//!
//! # Get the library with ID 0
//! curl -s localhost:8000/api/v0/libraries/0
//!
//! # Create a new peripheral from the library with ID 0
//! curl -s \
//!      --request POST \
//!      localhost:8000/api/v0/peripherals \
//!      --header "Content-Type: application/json" \
//!      --data '{"name":"foo","library_id":0}'
//!
//! # Create a new peripheral and override the default value of a pre-init attribute
//! curl -s \
//!      --request POST \
//!      localhost:8000/api/v0/peripherals \
//!      --header "Content-Type: application/json" \
//!      --data '{
//!          "name": "foo",
//!          "library_id": 0,
//!          "attributes": [
//!              {"id":0, "type":"double", "value": 999.99}
//!          ]
//!      }'
//!
//! # Get all the peripherals currently managed by the daemon
//! curl -s localhost:8000/api/v0/peripherals
//!
//! # Get the peripheral with ID 0
//! curl -s localhost:8000/api/v0/peripherals/0
//!
//! # Get the attributes of the peripheral with ID 0
//! curl -s localhost:8000/api/v0/peripherals/0/attributes
//!
//! # Get the attribute with ID 0 from the peripheral with ID 0
//! curl -s localhost:8000/api/v0/peripherals/0/attributes/0
//!
//! # Set the value of the attribute with ID 0 of the peripheral with ID 0
//! curl -s \
//!      --request PATCH \
//!      localhost:8000/api/v0/peripherals/0/attributes/0 \
//!      --header "Content-Type: application/json" \
//!      --data '{"type":"double","value":42}'
//! ```
//!
//! # Core components
//!
//!                +--------------------------------+
//!                |                                |
//!                |          Object Model          |
//!                |                                |      ^
//!                +--------------------------------+      |
//!             ----------------------------------------   |  User API
//!                +--------------------------------+      |
//!                |                                |      v
//!                |        REST Integration        |
//!                |                                |
//!                +--------------------------------+
//!             ----------------------------------------
//!                +--------------------------------+
//!                |                                |
//!                |           KPAL Daemon          |
//!                |                                |      ^
//!                +--------------------------------+      |
//!             ----------------------------------------   |  Plugin API
//!                +--------+  +--------+  +--------+      |
//!                |        |  |        |  |        |      v
//!                | Plugin |  | Plugin |  | Plugin |
//!                |        |  |        |  |        |
//!                +--------+  +--------+  +--------+
//!
//!
//! ## Object model
//!
//! The object model is the set of resources with which users interact. Currently, these resources
//! include:
//!
//! - **peripherals** Models of individual hardware peripherals
//! - **attributes** Attributes describe a peripheral. Users interact with peripherals by modifying
//! or reading their attributes.
//! - **libraries** The shared libraries that enable the plugin API
//!
//! ## Integrations
//!
//! Integrations are different implementations of the user API. Examples of possible integrations
//! include
//!
//! - a JSON REST API
//! - gRPC
//! - a C static library
//! - Python bindings
//!
//! Currently only a JSON REST integration is available, but the structure of KPAL makes it
//! relatively easy to add others.
//!
//! ## Daemon
//!
//! The KPAL daemon, or `kpald`, is a server that runs on the computer to which the peripherals are
//! connected. Users directly interact with the daemon through the user API. Each peripheral runs
//! inside its own thread which is spawned by a POST request to the user API. The daemon forwards
//! other user requests to each thread through the thread's dedicated channel. The threads
//! interpret the incoming requests and, in response, read and write data to individual plugins
//! through the plugin API using shared libraries.
//!
//! ### Plugins
//!
//! Plugins are the means by which peripherals are integrated into KPAL. A plugin uses a shared
//! library (a `.so` file on Linux) to communicate with the daemon. The common set of functions
//! that the library provides is the plugin API. Any programming language that can provide a C
//! language interface can be used to write a plugin library.
//!
//! A plugin combines the data that represents a peripheral's state with the functionality for
//! controlling the hardware device that is modeled by the peripheral.
pub mod constants;
pub mod init;
pub mod integrations;
pub mod models;
pub mod plugins;
